Feature/ucc listener run internally on controller and agent

morantr opened the pull request:
Replace cli_ucc_listener and wfa_ca module on the controller with controller and agent ucc listener threads.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3568,10 +3568,12 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
         LOG(TRACE) << "MASTER_CONNECTED";
 
+        master_socket = backhaul_manager_socket;
+        master_socket->setPeerMac(backhaul_params.controller_bridge_mac);
```
Are you sure this has no effect? When `wlan_settings.band_enabled` is 0, `master_socket` was never set, and now you always set it.

morantr replied:
The opposite, if the band enable is 0 the master socket was set immediately.
Basically, the cases in which the socket is not set are on all the cases we do slave_reset(), which we unset it there anyway. So if we didn't go to slave reset, we need to set it.

tomereli reviewed (changes requested):
Can you explain a bit more in the commit message?

tomereli reviewed (commented):
The first commit message is very informative, but it looks more like a brainstorming between you and yourself than a commit message. I would rephrase it to be something like:

> The ucc listener in the agent should ideally reside in the backhaul manager, since there should be only one. Unfortunately, doing so causes inconsistency on send_cmdu function which on the controller uses a socket, while in the backhaul manager is using the bus.
The son slave which runs the ucc listener is defined in the agent configuration file for the sake of simplicity.

However, I have several issues with this approach:

1. Having 2 different ports for UCC listeners for controller and agent is confusing. Why should we run both? We are either doing controller certification OR agent certification, never in the same time, right? So I would expect a single UCC listener which supports all types of commands. In this case I would still prefer the listener to be an external application which has access to the bus and can send CMDUs to the agent and controller - this way we can have a single termination point for all UCC commands.

2. Regarding "First I thought to implement it in the backhaul manager, but it caused
inconsistency on send cmdu function which on the controller, uses
a socket, and on the backhaul manager uses the bus, so I have decided to
implement it on son_slave." - I think the backhaul manager is the only logical place to run this thread, if we decide to go with this approach. Also you said yourself that we need to remove the mrouter (https://github.com/prplfoundation/prplMesh/issues/435#issuecomment-542319551) so why not do that and align `send_cmdu`?

tomereli reviewed (commented):
The second commit is impossible to review, it would have been better to do it gradually so we can see which parts are copy paste from the wfa_ca / ucc listener. Is there any way you can make it easier to review? Perhaps add comments where the new implementation was modified?

tomereli reviewed (changes requested):
Fourth commit - I assume that a followup commit fixes test_flows.sh? Note that it has to be part of this PR, otherwise it won't pass CI..

tomereli reviewed (changes requested):
8th commit - I think that from this point this deserves a separate PR, after fixing the test_flows.sh.

tomereli reviewed (commented):
"Modify backhaul manager constructor" commit - Based on the commit message, this should be a fixup commit to the one that added `agent_ucc_listener`

morantr commented:
1)
> The first commit message is very informative, but it looks more like a brainstorming between you and yourself than a commit message. I would rephrase it to be something like:
> 
> > The ucc listener in the agent should ideally reside in the backhaul manager, since there should be only one. Unfortunately, doing so causes inconsistency on send_cmdu function which on the controller uses a socket, while in the backhaul manager is using the bus.
> > The son slave which runs the ucc listener is defined in the agent configuration file for the sake of simplicity.

Ok, acceptable, I will change it when I'll rebase.

2)
> However, I have several issues with this approach:
> 
> 1. Having 2 different ports for UCC listeners for controller and agent is confusing. Why should we run both? We are either doing controller certification OR agent certification, never in the same time, right? So I would expect a single UCC listener which supports all types of commands. In this case I would still prefer the listener to be an external application which has access to the bus and can send CMDUs to the agent and controller - this way we can have a single termination point for all UCC commands.

You can use whatever port you wish, it is configurable on the controller/agent config file.
You can even shut the ucc_listener by setting the port number to zero.
But you can't use them both on the same port, because it is impossible to bind a specific port twice.
Anyway, If I remember correctly, the ucc uses different ports for Agent and Controller. Also, even if we would have use external ucc_listener we still would have to use 2 instances which listen to 2 different port because it is impossible to know which command is destined for for each entity.

3) 
> 2. Regarding "First I thought to implement it in the backhaul manager, but it caused
>    inconsistency on send cmdu function which on the controller, uses
>    a socket, and on the backhaul manager uses the bus, so I have decided to
>    implement it on son_slave." - I think the backhaul manager is the only logical place to run this thread, if we decide to go with this approach. Also you said yourself that we need to remove the mrouter ([#435 (comment)](https://github.com/prplfoundation/prplMesh/issues/435#issuecomment-542319551)) so why not do that and align `send_cmdu`?

You are right, but for that to happen, first need to do some refactoring - deleting mrouter, and turn the master thread to bus socket thread.

4)
> The second commit is impossible to review, it would have been better to do it gradually so we can see which parts are copy paste from the wfa_ca / ucc listener. Is there any way you can make it easier to review? Perhaps add comments where the new implementation was modified?

Basically I didn't change much. Most of the code is copy-past. The only changes I did is to remove all bml related code from ucc_listener parts , and on wfa_ca part, create use the newly defined virtual functions on the command handler instead of just use them directly on the file. 

5)
> Fourth commit - I assume that a followup commit fixes test_flows.sh? Note that it has to be part of this PR, otherwise it won't pass CI..

I know. I am waiting for @rmelotte to finish the python script which will simulate the UCC, and then I'll push a fixup for that commit that aligns test_flows.sh with the new script.

6)
> 8th commit - I think that from this point this deserves a separate PR, after fixing the test_flows.sh.

These commit stand on its own and can be part of this PR since they are related. Besides, I already did it here, and I don't think it will give us any benefit to move it now.

7)
> "Modify backhaul manager constructor" commit - Based on the commit message, this should be a fixup commit to the one that added `agent_ucc_listener`

Why? How is it related to it? This commit is something I "fixed" which is not related to the PR (except the reason I did it).

rmelotte commented:
@morantr I didn't want to add commits on your branch directly in case you need to rewrite history, so I pushed to a separate branch instead: `dev/test_flows_capi_commands`.
At the moment, only one test is failing: test higher layer data payload.
It fails with:
```
status,INVALID,errorCode,param name 'tlv_value' has value with invalid format
```
I will look into this tomorrow, but as you can see there seem to be an issue with one of the validations of the `tlv_value` field.

rmelotte commented:
I added [a commit](https://github.com/prplfoundation/prplMesh/commit/a29d3effefa5e189e88a21c0413622a12db6ab55) on `dev/test_flows_capi_commands` to fix the test that was failing. 
The issue was in test_flows.sh, not the validation.

arnout commented on common/beerocks/bcl/source/beerocks_ucc_listener.cpp:
```diff
@@ -0,0 +1,1027 @@
+ *     echo -n <command> | nc <bridge_ip_addr> <port>
+ */
+
+#include "../include/beerocks/bcl/beerocks_ucc_listener.h"
```
AFAICS the include directory is passed with a `-I` option, so `#include <beerocks/bcl/beerocks_ucc_listener.h>` should work.

(nitpick)

tomereli reviewed (commented):
I really like commit 162f9e2 and especially the elaborate commit message. Kudos.

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -0,0 +1,133 @@
+ */
+void agent_ucc_listener::clear_configuration()
+{
+    // TODO
```
Why leave it as TODO? And if so, please refer to an open task.

morantr replied:
done

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -0,0 +1,133 @@
+{
+    // On the agent side, the dest_alid is not really needed since the destination socket will
+    // always be the controller socket.
+    return *m_controller_sd;
```
Can't `m_controller_sd` be nullptr? this may result with segfault, won't it?
Also, didn't you mean to do `return m_controller_sd` as you return a `Socket *`?

morantr replied:
see https://github.com/prplfoundation/prplMesh/pull/452#discussion_r342985412

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -0,0 +1,133 @@
+std::shared_ptr<uint8_t> agent_ucc_listener::get_buffer_filled_with_cmdu()
+{
+    // Currently, no such buffer on agent side.
+    return std::shared_ptr<uint8_t>();
```
so why not return `nullptr`?

morantr replied:
It is the same thing.

tomereli replied:
`nullptr` would be more appropriate IMHO. This just raises questions for anyone who will bump into this code. Also, should add a TODO to the comment with reference to an open task if this is bound to be changed.

morantr replied:
By definition shared_ptr<something>() is pointing to nullptr. 😒 
Whatever I'll change it.
There is no need to add a TODO, since we might not need certification buffer.

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -0,0 +1,133 @@
+ */
+bool agent_ucc_listener::send_cmdu_to_destination(Socket *sd, ieee1905_1::CmduMessageTx &cmdu_tx)
+{
+    if (*m_controller_sd == nullptr) {
```
What type is `m_controller_sd`? is it `shared_ptr`? why do you need to dereference it?

morantr replied:
see https://github.com/prplfoundation/prplMesh/pull/452#discussion_r342985412

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -0,0 +1,133 @@
+        return false;
+    }
+
+    // TODO
```
Please refer to an open task in the TODO

morantr replied:
done

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.h:
```diff
@@ -0,0 +1,49 @@
+    const std::string &m_model;
+    const std::string &m_bridge_iface;
+    std::string m_bridge_mac;
+    Socket **const m_controller_sd;
```
Oh, now I see it.... Why is it `**`?? Why not a simple pointer or shared_ptr?

morantr replied:
It is pointer to pointer since the son_slave is manging the controller socket pointer and it can modify it. That way it is promised that if the son_slave will change it, the ucc_listener would "know" about it.

tomereli replied:
This should be a shared_ptr then, and the son_slave should notify the listener on changes to the socket instead of relying on the pointer to be changed "under your feet".

morantr replied:
I don't agree with you. IMO this is the most clean/transparent way to do it. The design of the ucc_listener is to have a shared memory with another thread. On the master thread is easy since it has the controller database which holds the relevant sockets is on another file. Here it is more problematic since the socket is on the son_slave and I can't create a circular #include.
As well this is not the purpose of shared_ptr.
shared_ptr purourse is ti hold and not free a memory block as long as it is used. This is not the case here. Even if I turned the pointer to shared_ptr I would still have to pass it as a pointer or reference to shared_ptr to the ucc listener.
As far as I can see these are the alternative solutions:
1. Instead of passing the pointer to the controller socket, passing a callback function (lambda) which will give the current controller socket. The only drawback in this solution is that every time the ucc listener will need the controller socket, it will have to call to the callback function.
2. Create a public function on the ucc_listener that updates the controller socket.
This is a worse idea since every place the son_slave will want to change the controller socket, it will have to call this function. And of course, if a developer will want to change the controller socket in the future, it will have to remember to call this function. This can easily can be forgotten and can lead eventually to bugs.

tomereli replied:
> I don't agree with you. IMO this is the most clean/transparent way to do it. The design of the ucc_listener is to have a shared memory with another thread. On the master thread is easy since it has the controller database which holds the relevant sockets is on another file. Here it is more problematic since the socket is on the son_slave and I can't create a circular #include.

Can't you just use `extern`? no need for circular `#include`.

morantr replied:
I'll tell you what, When I will create the unified agent process, I'll create for it a database class like the controller has. That way the agent class will be much cleaner.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -119,6 +119,17 @@ bool slave_thread::init()
         return false;
     }
 
+    if (config.ucc_listener_slave_hostap_iface == config.hostap_iface &&
```
Please also add a flag which disables UCC completely in both agent and controller since it has effect on the code.

morantr replied:
There is already. It is the port number. If the port is zero, it means disable. (look one line down).

tomereli replied:
So this should be mentioned in a comment in the configuration file if it isn't already.

morantr replied:
done

tomereli reviewed (changes requested):
Mostly nitpicks, apart from the socket handling by the UCC listener. It should be notified by the slave on any changes to the socket. After this we can merge it.

tomereli commented on agent/src/beerocks/slave/agent_ucc_listener.cpp:
```diff
@@ -83,7 +84,7 @@ Socket *agent_ucc_listener::get_dev_send_1905_destination_socket(const std::stri
 {
     // Currently, no such buffer on agent side.
-    return std::shared_ptr<uint8_t>();
+    return std::shared_ptr<uint8_t>(nullptr);
```
I meant just `return nullptr` .... anyway (nitpick)

tomereli reviewed (approved):
I'm approving already so not to block this. But if you can fix the remaining comments it would be great.

arnout commented on agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h:
```diff
@@ -29,9 +30,9 @@ namespace backhaul_manager {
-                std::string const_bh_slave_);
+    main_thread(const config_file::sConfigSlave &config,
+                const std::set<std::string> &slave_ap_ifaces_,
+                const std::set<std::string> &slave_sta_ifaces_);
```
Since you're storing references to the arguments in the main_thread object, I think there should be a (doxygen) comment that clarifies that the lifetime of those arguments must surpass that of the main_thread object itself.

arnout reviewed (approved):
This is an excellent pull request, with nice commit messages and of course good content as well. I really like how the UCC listener has been implemented.

I've got two nitpick comments so it can be merged as is.

I'm not actually so sure about the "controller: Change database cetification buffer return value type" commit. It doesn't explain why we need to do that. But then, I don't really understand when smart pointers should be passed by reference and when by value.

morantr committed:
controller/agent: Add new ucc_listener configuration fields

Add new ucc_listener configuration fields both to Controller and Agent
configurations. On Controller configuration add only "ucc_listener_port"
field.
On Agent configuration add also "ucc_listener_port" field, but add as
well 'vendor' and 'model' for version command, and
'ucc_listener_slave_hostap_iface' to decide on which son_slave to run
the ucc_listenrer.

First I thought to implement it in the backhaul manager, but it caused
inconsistency on send cmdu function which on the controller, uses
a socket, and on the backhaul manager uses the bus, so I have decided to
implement it on son_slave.
I had two options to decide on which son_slave to run it:
1. Implement it as a singleton class.
2. Explicitly set on configuration file on which son_slave to run it.

I have chosen to set it on the configuration file because it is more
simple IMO.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: bcl: Add ucc listener thread

Add ucc listener thread that is a combination of cli ucc listener and
wfa_ca controller module.
To make common CA_GET_VERSION command handle to work, add
"BEEROCKS_VERSION" macro to common cmake file.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Add controller_ucc_listener thread

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

rmelotte committed:
tools: tests: add a python script to send CAPI commands

Previously, the commands were directly sent to beerocks_cli.  With
wfa_ca moving away from beerocks_cli, there will be a dedicated socket
for the CAPI commands.

This commit adds a python script that makes it easier to send CAPI
commands to the listening socket, from the shell test
script (i.e. test_flows.sh).

The script expects three positional arguments: the host, the port and
the actual command. It connects to the target socket, send the
command, and waits until it receives a COMPLETE, INVALID, or ERROR
reply.
The replies are printed by the script as they are received.
If an INVALID or ERROR reply is received, the script ends with an
exception.
If the device replies with an unexpected string, the script also
ends with an exception.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
travis.yml: switch to python 3.7.

The Travis builds currently only use scripts that are python3
compatible.

As python2 will be deprecated soon anyway, this commit replaces it
with python3.

Should any later build fail because of the python version, they will
need to be fixed to use python3.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: test_flows: remove extra space from the payload

Previously the payload was built by adding a space then the gateway
mac address to the payload variable.  As the variable was initially
empty, the resulting payload contained an extra space at the
beginning.

This commit initializes the payload variable so that there is no space
at the beginning of the resulting payload. As the payload already
contains one mac address when entering the for loop, it has to run one
time less.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

morantr committed:
controller/tools: Use new controller ucc listener thread

controller changes:
Add a new controller ucc listener thread to the controller.
This change required to remove wfa_ca module from the controller to
prevent compilation error due to conflicts.

tools changes:
update tests_flows.sh to use the python script to send commands
test_flows.sh should be able to send CAPI commands to both the
controller and agents.

Instead of using beerocks_cli to send CAPI commands, this commit uses
the new "send_CAPI_command.py".

Three new functions have been added to get the container port, IP, and
to send the CAPI commands.

Existing calls to send_bml_command have to be updated to remove the
"bml_wfa_ca_controller" and give the container name instead. Also, the
eval call and additional quoting are no longer necessary and have been
removed.

The send_bml_command has been kept because it's still being used, to
send steering requests for example.

Unfortunately, it's not currently possible to retrieve the container
IP without knowing the bridge name, because the IP that docker assigns
itself is not the one that is actually used for the bridge. For this
reason, the bridge_name had to be hard-coded to "br-lan".

The changes in controller and tools are not in a separated commit, to
not break the commit.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

rmelotte committed:
tools: test_flows.sh: use check and check_error

Some tests were already using check() and check_error, but not
all. Also, some tests were using them, but did not return check_error,
which made the test always pass if the last command succeeded (even if
any previous ones failed).

With the new python script to send commands, it's now also possible to
check the output of the call (the script returns an error code if the
device replied with an error for example).

This commit adds check() calls where possible, and set and return
check_error accordingly. As a consequence, failing tests should no
longer be marked as "OK", even if the last command of the test succeeded.

Signed-off-by: Raphaël Mélotte <raphael.melotte@essensium.com>

morantr committed:
cli: Remove cli_ucc_listener from cli

Remove cli_ucc_listener from cli since it is not used anymore.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller/bml: Remove wfa_ca datapath from cli to the controller

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
common: bcl: Remove wfa_ca messages

Remove wfa_ca messages from tlvf since there are no longer used.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add agent_ucc_listener implementation

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Save controller bridge mac on controller socket

Preparative commit.
Save controller bridge mac on controller socket so it could be used
on send_cmdu_to_destination function on agent_ucc_listener.
Also, change the update of master_socket (aka socket to the
controller) to be in a single place.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Add agent_ucc_listener to son_slave

Add agent_ucc_listener to son_slave.
In this scope, add to son_slave lock and unlock calls to before and
after select, because it has a shared memory with the ucc_listener.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
agent: Modify backhaul manager constructor

While implementing agent_ucc_listener, the thought was to locate it on
the backhaul manager. Eventually, it was not implemented on the backhaul
manager, but some changes made during the implementation process,
on class constructor and initialization which are more efficient, and
makes the code cleaner, so I decided to push it.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>

morantr committed:
controller: Change database cetification buffer return value type

Removed the reference from the return value type of the certification
buffer allocate and get function.

Signed-off-by: Moran Shoeg <moranx.shoeg@intel.com>
